index.js

```diff
@@ -153,54 +153,87 @@ function SpotifyWebHelper(opts) {
 			});
 		});
 	};
-	this.generateSpotifyUrl = function (url) {
+	this.generateSpotifyUrl = function (url, port) {
+		port = typeof port === "undefined" ? localPort : port;
 		var protocol = 'https://';
-		if (localPort >= START_HTTP_PORT && localPort <= END_HTTP_PORT) {
+		if (port >= START_HTTP_PORT && port <= END_HTTP_PORT) {
 			protocol = 'http://';
 		}
-		return util.format('%s%s:%d%s', protocol, '127.0.0.1', localPort, url);
+		return util.format('%s%s:%d%s', protocol, '127.0.0.1', port, url);
 	};
 	this.getOauthToken = function () {
 		return new Promise(function (resolve, reject) {
 			getJSON({
 				url: 'http://open.spotify.com/token'
 			})
-			.then(function (res) {
-				resolve(res.t);
-			})
-			.catch(reject);
+				.then(function (res) {
+					resolve(res.t);
+				})
+				.catch(reject);
 		});
 	};
-  this.checkForError = status => {
-    if (!status.open_graph_state) {
-      this.player.emit('error', new Error('No user logged in'));
-      return true;
-    }
-    if (status.error) {
-      this.player.emit('error', new Error(status.error.message));
-      return true;
-    }
-    return false;
-  };
-	this.detectPort = function () {
+	this.checkForError = status => {
+		if (!status.open_graph_state) {
+			this.player.emit('error', new Error('No user logged in'));
+			return true;
+		}
+		if (status.error) {
+			this.player.emit('error', new Error(status.error.message));
+			return true;
+		}
+		return false;
+	};
+	this.checkPort = function(port) {
 		return getJSON({
-			url: this.generateSpotifyUrl('/service/version.json'),
+			url: this.generateSpotifyUrl('/service/version.json', port),
 			headers: ORIGIN_HEADER,
 			params: {
 				'service': 'remote'
 			}
-		})
-		.then(() => localPort)
-		.catch((err) => {
-			if (localPort === END_HTTP_PORT) {
+		}).then(() => {
+				return localPort;
+			})
+			.catch((err) => {
 				throw err;
-			} else if (localPort === END_HTTPS_PORT) {
-				localPort = START_HTTP_PORT;
-			} else {
-				++localPort;
-			}
-			return this.detectPort();
-		});
+			});
+	};
+	this.portCheckCount = 0;
+	this.foundPort = false;
+	this.checkPorts = function(start,end){
+		//console.log(start,end);
+		for(let i = start; i <= end; i++) {
+			this.portCheckCount++;
+			this.checkPort(i).then(() => {
+				if(this.foundPort) {
+					return;
+				}
+				this.foundPort = true;
+				localPort = i;
+				this.getOauthToken().then(oauthtoken => {
+						this.oauthtoken = oauthtoken;
+						console.log(oauthtoken);
+						return this.getCsrfToken();
+					})
+					.then(csrftoken => {
+						this.csrftoken = csrftoken;
+						console.log(csrftoken);
+						return getStatus();
+					})
+					.then(() => {
+						return listen();
+					})
+					.catch(err => this.player.emit('error', err));
+
+			}).catch((err) => {
+				if(--this.portCheckCount == 0) {
+					throw err;
+				}
+			});
+		}
+	};
+	this.detectPort = function () {
```

Comment:
Perhaps you want to stick with not clouding up the namespace and use return values for check ports.  You could have checkPorts return a promise which resolves to a port if it can find one in the range.  Promise.race would also work here if you did this
```javascript
Promise.race([
  this.checkPorts(START_HTTPS_PORT, END_HTTPS_PORT);
  this.checkPorts(START_HTTP_PORT, END_HTTP_PORT);
]).then(port => ...)
```
Of course this solution requires checkPort to gracefully fail without rejecting the promise or else race will reject as well.

Reply:
I've actually created an example using Promise.race that allows chaining the promises better and not crowding the namespace.
```javascript
// Return the first successful port 
this.checkPorts = function(startPort, endPort) {
  // The maximum time to wait for a success before failing
  const MAX_TIMEOUT = 5000;
  // Create an array of each port and race them
  return Promise.race([...Array(endPort - startPort).keys()].map(
    i => this.checkPort(startPort + i)
  ), new Promise((resolve, reject) => 
    setTimeout(() => reject('No port found in range'), MAX_TIMEOUT)
  ));
}
```
This solution means that `this.checkPort` needs to not throw when it doesn't find a port, but you can have the original promise chain at the bottom and clean it up a bit even if you're interested.

Reply:
I wasnt too sure what everything was returning and was just trying to make sure that it was still throwing the error and if the port wasnt found and still working as soon as it found the port. The fact that the race returns the promise like that makes this layout a lot neater and will deffo help me in future projects xD Thanks for the feedback :)

Reply:
Not sure if its just me messing up but the set timeout continues past the promise and causes it to crash due to unhandled promise rejections.